chia/clvm/node.py

```diff
@@ -0,0 +1,333 @@
+from typing import Optional, List, Dict, Tuple
+
+from chia.types.blockchain_format.sized_bytes import bytes32
+from chia.types.blockchain_format.coin import Coin
+from chia.types.blockchain_format.program import Program, SerializedProgram
+from chia.util.ints import uint64, uint32
+from chia.util.hash import std_hash
+from chia.util.errors import Err
+from chia.types.coin_record import CoinRecord
+from chia.types.spend_bundle import SpendBundle
+from chia.types.generator_types import BlockGenerator
+from chia.types.mempool_inclusion_status import MempoolInclusionStatus
+from chia.types.coin_solution import CoinSolution
+from chia.full_node.bundle_tools import simple_solution_generator
+from chia.full_node.mempool_manager import MempoolManager
+from chia.full_node.mempool_check_conditions import get_puzzle_and_solution_for_coin
+from chia.consensus.default_constants import DEFAULT_CONSTANTS
+from chia.consensus.coinbase import create_pool_coin, create_farmer_coin
+from chia.consensus.block_rewards import calculate_pool_reward, calculate_base_farmer_reward
+from chia.consensus.cost_calculator import NPCResult
+
+"""
+The purpose of this file is to provide a lightweight simulator for the testing of Chialisp smart contracts.
+
+The Node object uses actual MempoolManager and Mempool objects, while substituting the CoinStore, FullBlock, and
+BlockRecord objects for trimmed down versions.
+
+There is also a provided NodeClient object which implements many of the methods from chia.rpc.full_node_rpc_client
+and is designed so that you could test with it and then swap in a real rpc client that uses the same code you tested.
+"""
+
+
+class CoinStore:
+    def __init__(self):
+        self.coin_records: Dict[bytes32, CoinRecord] = {}
+
+    def add_coin_record(self, record: CoinRecord, allow_replace: bool = False):
```

Comment:
The standard pattern is called "CRUD" for "create, read, update, delete", and it's how most DBs work. In SQL, create corresponds to "insert", read to "select", update is "update" and delete is "delete". Separating out "create" and "update" into two functions might make sense here, rather than `allow_replace` (although this isn't really a big deal).

Reply:
That makes sense, but this is mimicking the full_node coin_store API which has this allow_replace feature.  If I'm going to keep this imitation of the CoinStore, I'd like to keep it as close to the original as possible